src/typegate/src/runtimes/typegraph/visibility.ts

```diff
@@ -0,0 +1,30 @@
+// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
+// SPDX-License-Identifier: MPL-2.0
+
+import { TypeGraphDS } from "../../typegraph/mod.ts";
+import { Runtime } from "../Runtime.ts";
+
+export class TypeVisibility {
+  visibilityCache: Map<number, boolean>;
+  constructor(
+    private tg: TypeGraphDS,
+    private denoRuntime: Runtime, // 
+    ) {
+    this.visibilityCache = new Map<number, boolean>();
+  }
+
+
+  // TODO:
+  // 
+  // Prepare visibility for each type in traversal order
+  // (How to reuse the same logic as the stage ones?? otherwise redo/refactor)
+  // If done, cache resolver results for the type idx
+  // 
```

Comment:
Note to later:
Should account for the path since types can be shared, visibilityCache should be used on a per type basis and better be renamed.